{% extends "home/base.html" %}
{% load staticfiles %}

{% block content %}
    <div class="row">
        
        <!-- News FIXME get this data from DB -->
        <div class='col-md-3'>
            <h4>Latest <a href='/news'>News</a></h4>
            <ul>
                <li>An <a href="http://dx.doi.org/10.1016/j.tips.2014.11.001">article</a>
                describing <a href='/documentation/numbering'>generic GPCR
                numbering schemes</a> published in <em>Trends in Pharmacological Sciences</em>.</li>
                <li>A new version of GPCRDB launched.</li>
                <li>An <a href="http://dx.doi.org/10.1016/j.ymeth.2014.09.009">article</a>
                describing <a href='/documentation/pharmacophore'>crystal structure
                pharmacophores</a> published in <em>Methods</em>.</li>
                <li>A <a href="/pages/poster">poster</a> describing GPCRDB presented at the GLISTEN conference in Budapest, Hungary.
                </li>
                <li>GPCRDB presented at the <a href="http://wcp2014.org/">World Congress of
                Pharmacology</a> in Cape Town, South Africa.</li>
                <li>GPCRDB <a href="http://dx.doi.org/10.1093/nar/gkt1255">published</a> in <em>Nucleic Acids Res.</em>
                </li>
            </ul>
        </div>

        <!-- Twitter feed -->
        <div class="col-md-3">
        <a class="twitter-timeline" href="https://twitter.com/gpcrdb" data-widget-id="645958683621498880">Tweets by @gpcrdb</a>
        <script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+"://platform.twitter.com/widgets.js";fjs.parentNode.insertBefore(js,fjs);}}(document,"script","twitter-wjs");</script>
        </div>

        <!-- Carousel -->
        <div class='col-md-6'>
            <div id="carousel-example-generic" class="carousel slide" data-ride="carousel">
                <!-- Indicators -->
                <ol class="carousel-indicators">
                    <li data-target="#carousel-example-generic" data-slide-to="0" class="active"></li>
                    <li data-target="#carousel-example-generic" data-slide-to="1"></li>
                    <li data-target="#carousel-example-generic" data-slide-to="2"></li>
                    <li data-target="#carousel-example-generic" data-slide-to="3"></li>
                </ol>

                <!-- Wrapper for slides -->
                <div class="carousel-inner" role="listbox">
                <div class="item active">
                    <img src="{% static 'home/carousel/diagrams.png' %}" alt="...">
                    <div class="carousel-caption">
                        <h2>Build diagrams</h2>
                        <p>Get helix box and/or snake plot diagrams with custom coloring for you publications.</p>
                    </div>
                </div>
                <div class="item">
                    <img src="{% static 'home/carousel/alignments.png' %}" alt="...">
                    <div class="carousel-caption">
                        <h2>Build alignments</h2>
                        <p>Select proteins to align individually or by family. Choose custom sequence segements to include in your alignment.</p>
                    </div>
                </div>
                <div class="item">
                    <img src="{% static 'home/carousel/trees.png' %}" alt="...">
                    <div class="carousel-caption">
                        <h2>Build phylogenetic trees</h2>
                        <p>Customise the sequence segments to use for generating the trees.</p>
                    </div>
                </div>
                <div class="item">
                    <img src="{% static 'home/carousel/sequencemotifs.png' %}" alt="...">
                    <div class="carousel-caption">
                        <h2>Search for sequence motifs</h2>
                        <p>Predict off-targets by specifying sequence motifs that include residue positions and binding features.</p>
                    </div>
                </div>

                <!-- Controls -->
                <a class="left carousel-control" href="#carousel-example-generic" role="button" data-slide="prev">
                    <span class="glyphicon glyphicon-chevron-left" aria-hidden="true"></span>
                    <span class="sr-only">Previous</span>
                </a>
                <a class="right carousel-control" href="#carousel-example-generic" role="button" data-slide="next">
                    <span class="glyphicon glyphicon-chevron-right" aria-hidden="true"></span>
                    <span class="sr-only">Next</span>
                </a>
            </div>
        </div>

    </div>

    <div class="row">
        <div class="col-md-12 well">
            <div class='col-md-1'>
                <img src="http://gpcrdb.org/img/logo/gpcrdb.png" style="height:80px;width:80px;">
            </div>
            <div class='col-md-11'>
                <p><strong>GPCRdb</strong> contains data, diagrams and web tools for G protein-coupled receptors (GPCRs). Users can
                browse all GPCR crystal structures and the largest collections of mutants. Diagrams can be produced and
                downloaded to illustrate receptor residues (snake-plot and helix box diagrams) and relationships
                (phylogenetic trees). Reference (crystal) structure-based sequence alignments take into account helix bulges
                and constrictions, display statistics of amino acid conservation and have been assigned generic residue
                numbering for equivalent residues in different receptors. For an overview read the GPCRdb <a href="http://gpcrdb.org/about/poster">poster</a>,
                <a href="/pages/citing">articles</a> or
                <a href="/documentation">documentation</a>.</p>
            </div>
        </div>
    </div>

    <div class="row">
        <div class='col-md-3 well'>
            <h4>Partner GPCR databases</h4>
            <p>
                <a href='http://www.guidetopharmacology.org'>GuideToPharmacology</a> (IUPHAR),
                <a href='http://www.cmbi.ru.nl/tinygrap/credits'>tinyGRAP</a> (Troms√∏, Norway),
                <a href='http://www.gpcr-okb.org/'>GPCR-OKB</a> (Weill Medical College),
                <a href='http://nava.liacs.nl/'>NaVa</a> (Leiden Uni.)
                and <a href='http://web.expasy.org/docs/swiss-prot_guideline.html'>Swiss-Prot</a>.
            </p>
        </div>

        <div class='col-md-3'>
            <p class="text-center"><a href='http://gloriamgroup.org'>University of Copenhagen</p>
            <p class="text-center"><img src='http://gpcrdb.org/img/logo/uni_cph.png' style="height:100px;width:100px;"></a></p>
        </div>

        <div class='col-md-3'>
            <p>Supported by <a href='http://glisten-gpcr.eu'>GLISTEN</a> (COST Action CM1207)</p>
            <p><a href='http://glisten-gpcr.eu'><img src='http://gpcrdb.org/img/logo/glisten.png' style="height:28px;width:200px;"></a></p>
        </div>

        <div class='col-md-3 well'>
            <h4 class="text-center"><a href='/about/servers'>Partner GPCR servers</a></h4>
            <p class="text-center">
                <a href='http://gpcrm.biomodellab.eu/'>GPCRM</a> (Uni. Warsaw),
                <a href='http://cheminfo.u-strasbg.fr/scPDB/'>scPDB</a> (Uni. Strasbourg),
                <a href='http://www.ssfa-7tmr.de/ssfe/'>GPCR-SSFE</a> (Leibniz-Institut, Berlin),
                <a href='http://molsim.sci.univr.it/cgi-bin/cona/begin.php'>GoMoDo</a> (Uni. Verona),
                <a href='http://gpcr-modsim.org/'>GPCR-ModSim</a> (Uppsala Uni.).
            </p>
        </div>
        </div>
{% endblock %}